usb/device/USBCDC_ECM/USBCDC_ECM.cpp

```diff
@@ -0,0 +1,496 @@
+
+    if (new_state == Configured) {
+        _flags.set(FLAG_CONNECT);
+        _flags.clear(FLAG_DISCONNECT);
+    } else {
+        _flags.set(FLAG_DISCONNECT);
+        _flags.clear(FLAG_CONNECT | FLAG_WRITE_DONE | FLAG_INT_DONE);
+    }
+}
+
+const uint8_t *USBCDC_ECM::device_desc()
+{
+    uint8_t ep0_size = endpoint_max_packet_size(0x00);
+    uint8_t deviceDescriptorTemp[] = {
+        DEVICE_DESCRIPTOR_LENGTH,       // bLength
+        DEVICE_DESCRIPTOR,              // bDescriptorType
+        0x00, 0x02,                     // bcdUSB 2.0
+        0x02,                           // bDeviceClass
+        0x00,                           // bDeviceSubClass
+        0x00,                           // bDeviceProtocol
+        ep0_size,                       // bMaxPacketSize0
+        (uint8_t)(LSB(vendor_id)),
+        (uint8_t)(MSB(vendor_id)),      // idVendor
+        (uint8_t)(LSB(product_id)),
+        (uint8_t)(MSB(product_id)),     // idProduct
+        (uint8_t)(LSB(product_release)),
+        (uint8_t)(MSB(product_release)),// bcdDevice
+        STRING_OFFSET_IMANUFACTURER,    // iManufacturer
+        STRING_OFFSET_IPRODUCT,         // iProduct
+        STRING_OFFSET_ISERIAL,          // iSerialNumber
+        0x01                            // bNumConfigurations
+    };
+    MBED_ASSERT(sizeof(deviceDescriptorTemp) == sizeof(deviceDescriptor));
+    memcpy(deviceDescriptor, deviceDescriptorTemp, sizeof(deviceDescriptor));
+    return deviceDescriptor;
+}
+
+const uint8_t *USBCDC_ECM::string_iproduct_desc()
+{
+    static const uint8_t string_iproduct_descriptor[] = {
+        26,
+        STRING_DESCRIPTOR,
+        'U', 0, 'S', 0, 'B', 0, ' ', 0, 'E', 0, 't', 0, 'h', 0, 'e', 0, 'r', 0, 'n', 0, 'e', 0, 't', 0
+    };
+    return string_iproduct_descriptor;
+}
+
+const uint8_t *USBCDC_ECM::string_iconfiguration_desc()
+{
+    static uint8_t string_imac_addr[26] = {0};
+    const char unicodes[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+                             'A', 'B', 'C', 'D', 'E', 'F'
+                            };
+    char mac[6];
+
+    mbed_mac_address(mac);
+
+    string_imac_addr[0] = 26;
+    string_imac_addr[1] = STRING_DESCRIPTOR;
+    /* Convert MAC address to USB CDC string format */
+    for (int i = 0; i < 6; i++) {
+        string_imac_addr[i * 4 + 2] = unicodes[mac[i] >> 4];
+        string_imac_addr[i * 4 + 4] = unicodes[mac[i] & 0xF];
+    }
+    return string_imac_addr;
+}
+
+const uint8_t *USBCDC_ECM::string_iserial_desc()
+{
+    static const uint8_t string_iserial_descriptor[] = {
+        26,
+        STRING_DESCRIPTOR,
+        '0', 0, '1', 0, '2', 0, '3', 0, '4', 0, '5', 0, '6', 0, '7', 0, '8', 0, '9', 0, 'A', 0, 'B', 0
+    };
+    return string_iserial_descriptor;
+}
+
+#define CONFIG_DESC_SIZE (9+9+5+5+13+7+9+9+7+7)
+
+const uint8_t *USBCDC_ECM::configuration_desc(uint8_t index)
+{
+    if (index != 0) {
+        return NULL;
+    }
+
+    static const uint8_t configDescriptor[] = {
```

Comment:
This array should probably be a member of the class rather than static. Also the local static initialization will cause mutexes to be used in interrupt context.